### PR TITLE
TableNG: Table column filter

### LINKED ISSUE #94758: Table: Header - Filtering
- Enable column filter
- Contains
- Expression
- Multi-select

Image

### PR DESCRIPTION
**Pull request by Develer**
**What is this feature?**

This PR introduces table column filter.

**Which issue(s) does this PR fix?**:

Fixes #94758 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: packages/grafana-ui/src/components/Table/TableNG/Filter/Filter.tsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+interface Props {
+  name: string;
+  rows: any[];
+  filter: any;
+  setFilter: (value: any) => void;
+  field?: Field;
+}
+
+export const Filter = ({ name, rows, filter, setFilter, field }: Props) => {
+  const filterValue = filter[name]?.filtered;
```
Comment (drew08t) [inline packages/grafana-ui/src/components/Table/TableNG/Filter/Filter.tsx:22]:
```suggestion
  if (!field || !field.config.custom?.filterable) {
    return null;
  }
  const filterValue = filter[name]?.filtered;
```

Is there a reason this was removed? I noticed it was commented out in a previous commit. Without it, we can't do field overrides for disabling filter for a column.

Comment (drew08t) [reply packages/grafana-ui/src/components/Table/TableNG/Filter/Filter.tsx:22]: In fact, enabling a filter using `field.config.custom.filterable` is not supported as well.

File: packages/grafana-ui/src/components/Table/TableNG/TableNG.tsx [hunk 2]
```diff
@@ -137,7 +148,7 @@ export function TableNG(props: TableNGProps) {
   const defaultRowHeight = getDefaultRowHeight();
   const defaultLineHeight = theme.typography.body.lineHeight * theme.typography.fontSize;
 
-  const HeaderCell: React.FC<HeaderCellProps> = ({ column, onSort, direction, justifyContent }) => {
+  const HeaderCell: React.FC<HeaderCellProps> = ({ column, field, onSort, direction, justifyContent }) => {
     const headerRef = useRef(null);
```
Comment (drew08t) [inline packages/grafana-ui/src/components/Table/TableNG/TableNG.tsx:152]:
```suggestion
    let filterableWithOverride = filterable;
    if (field.config.custom.filterable !== undefined) {
      filterableWithOverride = field.config.custom.filterable;
    }
    const headerRef = useRef(null);
```

Something like this and then update line 176 to use `filterableWithOverride`.

Comment (drew08t) [reply packages/grafana-ui/src/components/Table/TableNG/TableNG.tsx:152]: This only makes the UI support filter overrides, but doesn't go the full way to making it fully functional.

**Comment by drew08t** [comment]
In addition to the comments above, I was having trouble seeing persistence on expression values.

Table OG:
Nov-25-2024 18-48-07

Table NG:
Nov-25-2024 18-46-46

**Comment by drew08t** [comment]
Let's also make sure to capture the follow-up task for cross-column filter dependency if you haven't already.

File: packages/grafana-ui/src/components/Table/TableNG/Filter/Filter.tsx (hunk 1 above)
Comment (Develer) [reply packages/grafana-ui/src/components/Table/TableNG/Filter/Filter.tsx:22]: indeed, let me add it back

**Comment by Develer** [comment]
Here is a follow-up task on cross-column filter dependency:
https://github.com/grafana/grafana/issues/97045

**Comment by drew08t** [comment]
1. Persistence on expression / value is working as expected 🚀
2. Custom filter isn't quite working properly
    - If set to true globally, default filterable override (which is actually undefined, not false for some reason) will not override
        - We need to change the logic to simply compare `filterable` with `field.config.custom.filterable` and if they aren't equal, use `field.config.custom.filterable`
    - Custom filters that do show up aren't actually functional for me. Are you experiencing the same?

**Comment by drew08t** [comment]
I pushed a commit to make sure filteredRows updates when sortedRows changed.

One issue remaining however is footer calculations. Because footer calculations are done within `renderSummaryCell` inside of `mapFrameToDataGrid` they are dependent upon `rows` instead of `filteredRows`. Expected behavior is for footer calcs to be dependent upon the filtered content.

**Comment by Develer** [comment]
> I pushed a commit to make sure filteredRows updates when sortedRows changed.
> 
> One issue remaining however is footer calculations. Because footer calculations are done within `renderSummaryCell` inside of `mapFrameToDataGrid` they are dependent upon `rows` instead of `filteredRows`. Expected behavior is for footer calcs to be dependent upon the filtered content.

oh, I see, let me take a look at it

**Pull request merged**